File: server/douyin.py
"""Durable Douyin parse/download tasks backed by a bounded yt-dlp process."""
from __future__ import annotations

import json
import os
import re
import selectors
import shutil
import signal
import subprocess
import tempfile
import threading
import time
import uuid
from pathlib import Path
from urllib.parse import urlsplit, urlunsplit

from .errors import ApiError
from .security import validate_id
from .storage import JsonStore

HOSTS = {"douyin.com", "www.douyin.com", "m.douyin.com", "v.douyin.com", "iesdouyin.com", "www.iesdouyin.com"}
ACTIVE = {"queued", "running", "cancelling"}


def extract_url(text):
    if not isinstance(text, str) or not 1 <= len(text) <= 8192:
        raise ApiError(400, "invalid_link", "Provide a Douyin link or share text (maximum 8192 characters)")
    match = re.search(r'https://[^\s]+', text)
    try:
        url = urlsplit(match[0].rstrip(',.。，!?！？;；:)）]}》」\"\'') if match else "")
        if url.scheme != "https" or url.hostname not in HOSTS or url.username or url.password or url.port or not url.path:
            raise ValueError()
        return urlunsplit((url.scheme, url.netloc, url.path, url.query, ""))
    except ValueError:
        raise ApiError(400, "invalid_link", "Only public Douyin HTTPS links are supported") from None


class DouyinTasks:
    def __init__(self, config, assets, mutation_lock, media):
        self.config, self.assets, self.mutation_lock, self.media = config, assets, mutation_lock, media
        self.store = JsonStore(config.data_root / "metadata" / "douyin-tasks")
        self.lock = threading.RLock()
        self.events, self.threads = {}, {}
        self.slots = threading.BoundedSemaphore(2)
        self.stopping = False
        self.executable = os.environ.get("H3_STUDIO_YTDLP", "yt-dlp")
        self.cookie_file = os.environ.get("H3_STUDIO_DOUYIN_COOKIES", "")
        self.timeout = 600
        for task in self.store.list():
            if task["status"] in ACTIVE:
                self.update(task["id"], status="failed", error={"code": "interrupted", "message": "Server restarted; retry this task", "retryable": True})

    def capabilities(self):
        available = all(shutil.which(x) for x in (self.executable, "ffmpeg", "ffprobe"))
        return {"available": available, "provider": "yt-dlp", "cookie_configured": bool(self.cookie_file),
                "reason": "" if available else "Install yt-dlp, ffmpeg and ffprobe on the Studio server", "max_video_bytes": self.config.max_video_bytes}

    def submit(self, request):
        if set(request) - {"text", "mode", "quality", "request_id"}:
            raise ApiError(400, "invalid_parameter", "Expected text, mode, quality and optional request_id")
        url = extract_url(request.get("text"))
        mode, quality = request.get("mode", "download"), request.get("quality", "best")
        if mode not in {"parse", "download"} or quality not in {"best", "1080", "720"}:
            raise ApiError(400, "invalid_parameter", "Invalid download mode or quality")
        request_id = validate_id(request.get("request_id", uuid.uuid4().hex), "request_id")
        with self.lock:
            for old in self.store.list():
                if old.get("request_id") == request_id:
                    if (old["url"], old["mode"], old["quality"]) != (url, mode, quality):
                        raise ApiError(409, "idempotency_conflict", "request_id already used with another request")
                    return self.public(old)
            if self.stopping or len(self.events) >= 8:
                raise ApiError(429, "download_queue_full", "Download queue is full; retry later")
            if not self.capabilities()["available"]:
                raise ApiError(503, "dependency_missing", self.capabilities()["reason"])
            now = time.time()
            task = {"id": uuid.uuid4().hex, "request_id": request_id, "url": url, "mode": mode, "quality": quality,
                    "status": "queued", "stage": "queued", "progress": 0, "created_at": now, "updated_at": now}
            self.store.put(task["id"], task)
            event = self.events[task["id"]] = threading.Event()
            worker = self.threads[task["id"]] = threading.Thread(target=self.run, args=(task, event), daemon=True)
            worker.start()
            return self.public(task)

    def update(self, task_id, **values):
        with self.lock:
            task = self.store.get(task_id)
            task.update(values, updated_at=time.time())
            self.store.put(task_id, task)
            return task

    def public(self, task):
        value = dict(task)
        value["task_id"] = task["id"]
        if task.get("asset_id"):
            try:
                value["asset"] = self.assets.public_metadata(self.assets.get(task["asset_id"]))
            except ApiError:
                value["asset_missing"] = True
        return value

    def get(self, task_id):
        return self.public(self.store.get(validate_id(task_id, "task_id")))

    def list(self):
        return {"tasks": [self.public(t) for t in self.store.list()[:100]]}

    def cancel(self, task_id):
        with self.lock:
            task = self.store.get(validate_id(task_id, "task_id"))
            if task["status"] in ACTIVE:
                # Asset import is the commit point; never report cancellation after importing.
                if task["stage"] == "importing":
                    raise ApiError(409, "import_in_progress", "Asset import is finishing; wait for completion")
                event = self.events.get(task_id)
                if event:
                    event.set()
                task = self.update(task_id, status="cancelling")
            return self.public(task)

    def retry(self, task_id):
        task = self.store.get(validate_id(task_id, "task_id"))
        if task["status"] not in {"failed", "canceled"}:
            raise ApiError(409, "task_not_retryable", "Only failed or canceled tasks can be retried")
        return self.submit({"text": task["url"], "mode": task["mode"], "quality": task["quality"]})

    def stop(self):
        with self.lock:
            self.stopping = True
            for event in self.events.values():
                event.set()
            workers = list(self.threads.values())
        for worker in workers:
            worker.join(timeout=2)

    def command(self, args, event, work, progress=None):
        base = [self.executable, "--ignore-config", "--no-plugin-dirs", "--no-playlist", "--no-colors",
                "--socket-timeout", "20", "--retries", "2", "--extractor-retries", "2"]
        if self.cookie_file:
            # yt-dlp may update its cookie jar; use a private disposable copy.
            cookie = work / "cookies.txt"
            if not cookie.exists():
                shutil.copyfile(self.cookie_file, cookie)
                cookie.chmod(0o600)
            base += ["--cookies", str(cookie)]
        process = subprocess.Popen(base + args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, start_new_session=True)
        buffers = {"out": bytearray(), "err": bytearray()}
        pending = bytearray()
        deadline = time.monotonic() + self.timeout
        try:
            with selectors.DefaultSelector() as selector:
                selector.register(process.stdout, selectors.EVENT_READ, "out")
                selector.register(process.stderr, selectors.EVENT_READ, "err")
                while selector.get_map():
                    if event.is_set():
                        raise ApiError(409, "canceled", "Download canceled")
                    if time.monotonic() > deadline:
                        raise ApiError(504, "download_timeout", "Download timed out; retry later")
                    if sum(p.stat().st_size for p in work.iterdir() if p.is_file()) > self.config.max_video_bytes * 2 + 8 * 1024 * 1024:
                        raise ApiError(413, "download_too_large", "Download exceeds the video size limit")
                    for key, _ in selector.select(0.2):
                        chunk = os.read(key.fileobj.fileno(), 65536)
                        if not chunk:
                            selector.unregister(key.fileobj)
                            continue
                        buffers[key.data].extend(chunk)
                        if len(buffers[key.data]) > 8 * 1024 * 1024:
                            raise ApiError(502, "extractor_output_limit", "Extractor output exceeded its limit")
                        if key.data == "out" and progress:
                            pending.extend(chunk)
                            while b"\n" in pending:
                                line, _, rest = pending.partition(b"\n")
                                pending = bytearray(rest)
                                progress(line.decode("utf-8", "replace"))
            process.wait(timeout=5)
            if process.returncode:
                message = buffers["err"].decode("utf-8", "replace").lower()
                if "http error 429" in message or "too many requests" in message:
                    raise ApiError(429, "rate_limited", "Douyin is limiting requests. Stop repeated attempts and retry later.")
                if any(word in message for word in ("cookies", "sign in", "login", "captcha")):
                    raise ApiError(422, "cookie_refresh_required", "Douyin rejected the request. This does not prove that the session expired; verify the video in the browser and retry later.")
                if "http error 403" in message or "403 forbidden" in message:
                    raise ApiError(403, "access_restricted", "Douyin refused video access (HTTP 403). Verify the video in the browser and retry later.")
                raise ApiError(502, "extractor_failed", "Douyin extraction failed. Check the link, network and yt-dlp version, then retry.")
            return buffers["out"].decode("utf-8", "replace")
        finally:
            if process.poll() is None:
                os.killpg(process.pid, signal.SIGKILL)
            process.wait()
            process.stdout.close()
            process.stderr.close()

    def run(self, task, event):
        acquired = False
        task_id = task["id"]
        try:
            while not acquired:
                if event.is_set():
                    raise ApiError(409, "canceled", "Download canceled")
                acquired = self.slots.acquire(timeout=0.2)
            with tempfile.TemporaryDirectory(prefix="douyin-", dir=self.config.data_root / "tmp") as directory:
                work = Path(directory)
                self.update(task_id, status="running", stage="parsing")
                raw = json.loads(self.command(["--skip-download", "--dump-single-json", "--", task["url"]], event, work))
                if not raw.get("id") or raw.get("_type") == "playlist":
                    raise ApiError(422, "invalid_response", "Expected one Douyin video")
                metadata = {k: raw[k] for k in ("id", "title", "uploader", "duration", "width", "height") if k in raw}
                self.update(task_id, metadata=metadata)
                if task["mode"] == "parse":
                    with self.lock:
                        if event.is_set():
                            raise ApiError(409, "canceled", "Parsing canceled")
                        self.update(task_id, status="completed", stage="completed", progress=100)
                    return
                self.update(task_id, stage="downloading")
                last_progress = -1

                def progress(line):
                    nonlocal last_progress
                    if line.startswith("progress:"):
                        try:
                            p = json.loads(line[9:])
                            total = p.get("total_bytes") or p.get("total_bytes_estimate") or 0
                            percent = min(95, int(95 * p.get("downloaded_bytes", 0) / total)) if total else 0
                            if percent > last_progress:
                                last_progress = percent
                                self.update(task_id, progress=percent)
                        except (ValueError, TypeError, ZeroDivisionError):
                            pass
                fmt = "bv*+ba/b" if task["quality"] == "best" else f"bv*[height<={task['quality']}]+ba/b[height<={task['quality']}]"
                self.command(["--newline", "--progress", "--progress-template", "download:progress:%(progress)j", "--no-simulate",
                              "--max-filesize", str(self.config.max_video_bytes), "--merge-output-format", "mp4",
                              "-f", fmt, "-o", str(work / "video.%(ext)s"), "--", task["url"]], event, work, progress)
                files = [p for p in work.glob("video.*") if p.suffix in {".mp4", ".webm", ".mov", ".mkv"}]
                if len(files) != 1:
                    raise ApiError(422, "download_missing", "No complete video was downloaded (check the size limit)")
                with self.lock:
                    if event.is_set():
                        raise ApiError(409, "canceled", "Download canceled")
                    self.update(task_id, stage="importing", progress=97)
                source = files[0]
                digest = self.assets.hash_file(source)
                with self.mutation_lock:
                    asset = self.assets.find_library_duplicate(digest, requested_kind="video")
                    reused = asset is not None
                    if asset is None:
                        if self.assets.used_bytes() + self.media.quota_bytes() + source.stat().st_size > self.config.max_asset_storage_bytes:
                            raise ApiError(507, "asset_quota", "Asset storage quota exceeded")
                        asset = self.assets.import_file(source, original_filename=f"douyin-{raw['id']}{source.suffix}", requested_kind="video")
                        if self.assets.used_bytes() + self.media.quota_bytes() > self.config.max_asset_storage_bytes:
                            self.assets.delete(asset["id"])
                            raise ApiError(507, "asset_quota", "Normalized video exceeds storage quota")
                    self.update(task_id, status="completed", stage="completed", progress=100, asset_id=asset["id"], reused=reused)
        except ApiError as error:
            canceled = error.code == "canceled" or (event.is_set() and self.store.get(task_id).get("stage") != "importing")
            self.update(task_id, status="canceled" if canceled else "failed", error={"code": "canceled" if canceled else error.code, "message": "Download canceled" if canceled else error.message, "retryable": not canceled})
        except Exception:
            self.update(task_id, status="failed", error={"code": "download_failed", "message": "Download or asset import failed; check dependencies and retry", "retryable": True})
        finally:
            if acquired:
                self.slots.release()
            with self.lock:
                self.events.pop(task_id, None)
                self.threads.pop(task_id, None)

File: cli/internal/douyin/douyin.go
package douyin

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"
)

const maxProcessOutput = 8 << 20

var linkPattern = regexp.MustCompile(`https://[^\s]+`)

type Error struct {
	Code      string
	Message   string
	Retryable bool
	Cause     error
}

func (e *Error) Error() string { return e.Message }
func (e *Error) Unwrap() error { return e.Cause }

type Format struct {
	ID       string `json:"id,omitempty"`
	Ext      string `json:"ext,omitempty"`
	Width    int    `json:"width,omitempty"`
	Height   int    `json:"height,omitempty"`
	VCodec   string `json:"vcodec,omitempty"`
	ACodec   string `json:"acodec,omitempty"`
	FileSize int64  `json:"filesize,omitempty"`
}

type Metadata struct {
	ID         string   `json:"id"`
	Title      string   `json:"title,omitempty"`
	Uploader   string   `json:"uploader,omitempty"`
	Duration   float64  `json:"duration,omitempty"`
	Thumbnail  string   `json:"thumbnail,omitempty"`
	WebpageURL string   `json:"webpage_url,omitempty"`
	Width      int      `json:"width,omitempty"`
	Height     int      `json:"height,omitempty"`
	Ext        string   `json:"ext,omitempty"`
	Formats    []Format `json:"formats,omitempty"`
}

type DownloadResult struct {
	Path   string `json:"path,omitempty"`
	Size   int64  `json:"size"`
	SHA256 string `json:"sha256,omitempty"`
}

type Extractor interface {
	Parse(context.Context, string) (Metadata, error)
	Download(context.Context, string, string, bool) (DownloadResult, error)
}

type CommandResult struct {
	Stdout string
	Stderr string
}

type CommandRunner func(context.Context, string, ...string) (CommandResult, error)

type Config struct {
	Executable         string
	CookiesFromBrowser string
	Timeout            time.Duration
	Run                CommandRunner
}

type Client struct {
	executable string
	browser    string
	timeout    time.Duration
	run        CommandRunner
}

func New(config Config) (*Client, error) {
	executable, err := resolveExecutable(config.Executable)
	if err != nil {
		return nil, err
	}
	if err := validateBrowser(config.CookiesFromBrowser); err != nil {
		return nil, err
	}
	if config.Timeout <= 0 {
		config.Timeout = 90 * time.Second
	}
	run := config.Run
	if run == nil {
		run = runCommand
	}
	return &Client{executable: executable, browser: config.CookiesFromBrowser, timeout: config.Timeout, run: run}, nil
}

func ExtractURL(text string) (string, error) {
	raw := linkPattern.FindString(strings.TrimSpace(text))
	raw = strings.TrimRight(raw, `,.。，!?！？;；:)）]}》」"'`)
	if raw == "" {
		return "", &Error{Code: "invalid_link", Message: "no HTTPS link was found in the supplied text"}
	}
	parsed, err := url.Parse(raw)
	if err != nil || parsed.Scheme != "https" || parsed.User != nil || parsed.Port() != "" {
		return "", &Error{Code: "invalid_link", Message: "the Douyin link is invalid"}
	}
	host := strings.ToLower(parsed.Hostname())
	allowed := map[string]bool{
		"douyin.com": true, "www.douyin.com": true, "m.douyin.com": true,
		"v.douyin.com": true, "iesdouyin.com": true, "www.iesdouyin.com": true,
	}
	if !allowed[host] {
		return "", &Error{Code: "invalid_link", Message: "only public douyin.com and iesdouyin.com links are accepted"}
	}
	parsed.Fragment = ""
	return parsed.String(), nil
}

func (c *Client) Parse(ctx context.Context, text string) (Metadata, error) {
	link, err := ExtractURL(text)
	if err != nil {
		return Metadata{}, err
	}
	ctx, cancel := context.WithTimeout(ctx, c.timeout)
	defer cancel()
	args := c.baseArgs()
	args = append(args, "--skip-download", "--dump-single-json", "--", link)
	result, runErr := c.run(ctx, c.executable, args...)
	if runErr != nil {
		return Metadata{}, classifyProcessError(ctx, result.Stderr, runErr)
	}
	var raw struct {
		ID         string  `json:"id"`
		Title      string  `json:"title"`
		Uploader   string  `json:"uploader"`
		Thumbnail  string  `json:"thumbnail"`
		WebpageURL string  `json:"webpage_url"`
		Ext        string  `json:"ext"`
		Duration   float64 `json:"duration"`
		Width      int     `json:"width"`
		Height     int     `json:"height"`
		Formats    []struct {
			FormatID string `json:"format_id"`
			Ext      string `json:"ext"`
			VCodec   string `json:"vcodec"`
			ACodec   string `json:"acodec"`
			Width    int    `json:"width"`
			Height   int    `json:"height"`
			FileSize int64  `json:"filesize"`
		}
	}
	if err := json.Unmarshal([]byte(result.Stdout), &raw); err != nil {
		return Metadata{}, &Error{Code: "invalid_response", Message: "yt-dlp returned invalid metadata", Cause: err}
	}
	if raw.ID == "" {
		return Metadata{}, &Error{Code: "invalid_response", Message: "yt-dlp metadata did not include a video id"}
	}
	metadata := Metadata{
		ID: raw.ID, Title: raw.Title, Uploader: raw.Uploader, Duration: raw.Duration,
		Thumbnail: raw.Thumbnail, WebpageURL: raw.WebpageURL, Width: raw.Width, Height: raw.Height, Ext: raw.Ext,
	}
	for _, item := range raw.Formats {
		if item.VCodec == "none" {
			continue
		}
		metadata.Formats = append(metadata.Formats, Format{
			ID: item.FormatID, Ext: item.Ext, Width: item.Width, Height: item.Height,
			VCodec: item.VCodec, ACodec: item.ACodec, FileSize: item.FileSize,
		})
	}
	return metadata, nil
}

func (c *Client) Download(ctx context.Context, text, outputTemplate string, force bool) (DownloadResult, error) {
	link, err := ExtractURL(text)
	if err != nil {
		return DownloadResult{}, err
	}
	if strings.TrimSpace(outputTemplate) == "" || strings.ContainsAny(outputTemplate, "\x00\r\n") {
		return DownloadResult{}, &Error{Code: "invalid_output", Message: "an output path or template is required"}
	}
	ctx, cancel := context.WithTimeout(ctx, c.timeout)
	defer cancel()
	args := c.baseArgs()
	args = append(args, "--no-part", "--print", "after_move:filepath", "-o", outputTemplate)
	if force {
		args = append(args, "--force-overwrites")
	}
	args = append(args, "--", link)
	result, runErr := c.run(ctx, c.executable, args...)
	if runErr != nil {
		return DownloadResult{}, classifyProcessError(ctx, result.Stderr, runErr)
	}
	lines := strings.Split(strings.TrimSpace(result.Stdout), "\n")
	path := strings.TrimSpace(lines[len(lines)-1])
	if path == "" {
		return DownloadResult{}, &Error{Code: "invalid_response", Message: "yt-dlp did not report the downloaded file path"}
	}
	abs, err := filepath.Abs(path)
	if err != nil {
		return DownloadResult{}, &Error{Code: "invalid_response", Message: "yt-dlp returned an invalid file path", Cause: err}
	}
	info, err := os.Stat(abs)
	if err != nil || !info.Mode().IsRegular() {
		return DownloadResult{}, &Error{Code: "download_missing", Message: "yt-dlp completed without creating a regular media file", Cause: err}
	}
	digest, err := fileSHA256(abs)
	if err != nil {
		return DownloadResult{}, &Error{Code: "download_hash_failed", Message: "downloaded media could not be hashed", Cause: err}
	}
	return DownloadResult{Path: abs, Size: info.Size(), SHA256: digest}, nil
}

func (c *Client) baseArgs() []string {
	args := []string{
		"--ignore-config", "--quiet", "--no-warnings", "--no-playlist",
		"--socket-timeout", strconv.Itoa(max(10, int(c.timeout.Seconds()/3))),
		"--retries", "3", "--extractor-retries", "3", "--retry-sleep", "extractor:1",
	}
	if c.browser != "" {
		args = append(args, "--cookies-from-browser", c.browser)
	}
	return args
}

func resolveExecutable(explicit string) (string, error) {
	candidate := strings.TrimSpace(explicit)
	if candidate == "" {
		candidate = strings.TrimSpace(os.Getenv("H3CTL_YTDLP"))
	}
	if candidate == "" {
		candidate = "yt-dlp"
	}
	path, err := exec.LookPath(candidate)
	if err != nil {
		return "", &Error{Code: "dependency_missing", Message: "yt-dlp was not found; install it or pass --yt-dlp PATH", Cause: err}
	}
	return path, nil
}

func validateBrowser(value string) error {
	if len(value) > 256 || strings.ContainsAny(value, "\x00\r\n") {
		return &Error{Code: "invalid_browser", Message: "--cookies-from-browser contains invalid characters"}
	}
	return nil
}

func runCommand(ctx context.Context, executable string, args ...string) (CommandResult, error) {
	cmd := exec.CommandContext(ctx, executable, args...)
	var stdout, stderr cappedBuffer
	stdout.limit, stderr.limit = maxProcessOutput, maxProcessOutput
	cmd.Stdout, cmd.Stderr = &stdout, &stderr
	err := cmd.Run()
	if stdout.exceeded || stderr.exceeded {
		return CommandResult{Stdout: stdout.String(), Stderr: stderr.String()}, &Error{Code: "output_too_large", Message: "yt-dlp produced too much output"}
	}
	return CommandResult{Stdout: stdout.String(), Stderr: stderr.String()}, err
}

func classifyProcessError(ctx context.Context, stderr string, err error) error {
	lower := strings.ToLower(stderr)
	switch {
	case errors.Is(ctx.Err(), context.DeadlineExceeded):
		return &Error{Code: "timeout", Message: "Douyin parsing timed out", Retryable: true, Cause: err}
	case strings.Contains(lower, "http error 429") || strings.Contains(lower, "too many requests"):
		return &Error{Code: "rate_limited", Message: "Douyin is limiting requests. Stop repeated attempts, verify the video in the selected browser, and retry later", Retryable: true, Cause: err}
	case strings.Contains(lower, "fresh cookies") || strings.Contains(lower, "cookies are needed") || strings.Contains(lower, "sign in"):
		return &Error{Code: "cookie_refresh_required", Message: "Douyin rejected the request. Verify that the video plays in the selected browser, then retry later; this message does not prove that your cookies expired", Retryable: true, Cause: err}
	case strings.Contains(lower, "http error 403") || strings.Contains(lower, "403 forbidden"):
		return &Error{Code: "access_restricted", Message: "Douyin refused video access (HTTP 403). Verify the video in the selected browser and retry later; the response does not identify whether the account, session, or request was restricted", Retryable: true, Cause: err}
	case strings.Contains(lower, "unsupported url") || strings.Contains(lower, "invalid url"):
		return &Error{Code: "invalid_link", Message: "the supplied Douyin link could not be parsed", Cause: err}
	default:
		message := strings.TrimSpace(stderr)
		if message == "" {
			message = "yt-dlp failed to parse the Douyin link"
		}
		if len(message) > 1000 {
			message = message[:1000]
		}
		return &Error{Code: "extractor_failed", Message: message, Retryable: true, Cause: err}
	}
}

type cappedBuffer struct {
	buffer   bytes.Buffer
	limit    int
	exceeded bool
}

func (w *cappedBuffer) Write(value []byte) (int, error) {
	original := len(value)
	remaining := w.limit - w.buffer.Len()
	if remaining <= 0 {
		w.exceeded = true
		return original, nil
	}
	if len(value) > remaining {
		value = value[:remaining]
		w.exceeded = true
	}
	_, _ = w.buffer.Write(value)
	return original, nil
}

func (w *cappedBuffer) String() string { return w.buffer.String() }

func fileSHA256(path string) (string, error) {
	file, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer file.Close()
	digest := sha256.New()
	if _, err := io.Copy(digest, file); err != nil {
		return "", err
	}
	return fmt.Sprintf("%x", digest.Sum(nil)), nil
}

func MapError(err error) *Error {
	var typed *Error
	if errors.As(err, &typed) {
		return typed
	}
	return &Error{Code: "internal_error", Message: fmt.Sprintf("Douyin operation failed: %v", err), Cause: err}
}
